STANDARDS.md

```diff
@@ -1,3 +1,1269 @@
+public interface a module provides. It gives us stability guarantees (namely, we
+know that we can change things that aren't exported without breaking downstream
+code at compile time), and tells us where to go looking first when inspecting or
+learning the module. Additionally, it means there is less change that
+implementation details 'leak' out of the module due to mistakes on the part of
+developers, especially those who may not be familiar with the code in question.
+
+Allowing wildcard _exports_ while disallowing wildcard _imports_ is justified on
+grounds of locality of information. Seeing a wildcard import of all a type's
+data constructors or fields doesn't tell us what these _are_ without looking up
+the module from where they are exported; having such an import be explicit
+reduces how much searching we have to do, as well as telling us clearly where
+certain identifiers come from. However, if we are reading an export list, we
+have the type definition in the same file we're already looking at, making it
+fairly easy to check.
+
+Re-exports are a useful feature, allowing us to have 'internal' versus
+'external' modules, which make for useful interface stability guarantees.
+However, _transitive_ re-exports are an unnecessary obfuscation, both for people
+familiarizing themselves with the code, and people trying to use the code (if
+it's a library). Having to follow a daisy chain of multiple re-exports to get to
+a definition is tedious, and ultimately not necessary: the only distinction
+required are 'internal' modules and 'external' modules, where 'external' modules
+import 'internal' modules, and then selectively re-export. The transitive
+re-export issue counts _double_ when dealing with identifiers from outside the
+project: in addition to being even more tedious, this is arguably an abstraction
+boundary violation. We forbid `module` re-exports on the same grounds as we
+restrict imports: without knowing exactly what we're re-exporting, it makes it
+very difficult to see what exact public interface we are presenting to the
+world.
+
+### ``LANGUAGE`` pragmata
+
+The following pragmata MUST be enabled at the project level (that is, in the
+Cabal file, in the `default-extensions` section):
+
+* ``BangPatterns``
+* ``BinaryLiterals``
+* ``ConstraintKinds``
+* ``DataKinds``
+* ``DeriveTraversable``
+* ``DerivingVia``
+* ``EmptyCase``
+* ``FlexibleContexts``
+* ``FlexibleInstances``
+* ``GeneralisedNewtypeDeriving``
+* ``HexFloatLiterals``
+* ``InstanceSigs``
+* ``KindSignatures``
+* ``LambdaCase``
+* ``MultiParamTypeClasses``
+* ``NumericUnderscores``
+* ``OverloadedStrings``
+* ``ScopedTypeVariables``
+* ``StandaloneDeriving``
+* ``TupleSections``
+* ``TypeOperators``
+
+All other `LANGUAGE` pragmata MUST be enabled per-file. All `LANGUAGE` pragmata
+MUST be placed at the top of the source file, written as `{-# LANGUAGE
+PragmaName #-}`, with one `LANGUAGE` pragma per line.
+
```

Comment:
There ought to be some sensible wiggle-room here, too.

Reply:
Can you give an example here?